collectors/builtin/spark.py

```diff
@@ -0,0 +1,373 @@
+# [
+#     {
+#         "status": "COMPLETE",
+#         "stageId": 0,
+#         "attemptId": 0,
+#         "numActiveTasks": 100,
+#         "numCompleteTasks": 101,
+#         "numFailedTasks": 102,
+#         "executorRunTime": 103,
+#         "inputBytes": 104,
+#         "inputRecords": 105,
+#         "outputBytes": 106,
+#         "outputRecords": 107,
+#         "shuffleReadBytes": 108,
+#         "shuffleReadRecords": 109,
+#         "shuffleWriteBytes": 110,
+#         "shuffleWriteRecords": 111,
+#         "memoryBytesSpilled": 112,
+#         "diskBytesSpilled": 113,
+#         "name": "reduceByKey at <stdin>:1",
+#         "details": "",
+#         "schedulingPool": "default",
+#         "accumulatorUpdates": []
+#     }
+# ]
+STAGE_METRICS = [
+    'numActiveTasks',
+    'numCompleteTasks',
+    'numFailedTasks',
+    'executorRunTime',
+    'inputBytes',
+    'inputRecords',
+    'outputBytes',
+    'outputRecords',
+    'shuffleReadBytes',
+    'shuffleReadRecords',
+    'shuffleWriteBytes',
+    'shuffleWriteRecords',
+    'memoryBytesSpilled',
+    'diskBytesSpilled'
+]
+
+# spark executor metric
+# [
+#     {
+#         "id": "driver",
+#         "hostPort": "10.0.2.15:33870",
+#         "rddBlocks": 99,
+#         "memoryUsed": 98,
+#         "diskUsed": 97,
+#         "activeTasks": 96,
+#         "failedTasks": 95,
+#         "completedTasks": 94,
+#         "totalTasks": 93,
+#         "totalDuration": 92,
+#         "totalInputBytes": 91,
+#         "totalShuffleRead": 90,
+#         "totalShuffleWrite": 89,
+#         "maxMemory": 278019440,
+#         "executorLogs": {}
+#     }
+# ]
+EXECUTOR_METRICS = [
+    'rddBlocks',
+    'memoryUsed',
+    'diskUsed',
+    'activeTasks',
+    'failedTasks',
+    'completedTasks',
+    'totalTasks',
+    'totalDuration',
+    'totalInputBytes',
+    'totalShuffleRead',
+    'totalShuffleWrite',
+    'maxMemory'
+]
+
+# spacke rdd metric
+# [
+#     {
+#         "id": 6,
+#         "name": "PythonRDD",
+#         "numPartitions": 2,
+#         "numCachedPartitions": 2,
+#         "storageLevel": "Memory Serialized 1x Replicated",
+#         "memoryUsed": 284,
+#         "diskUsed": 0
+#     }
+# ]
+RDD_METRCIS = [
+    'numPartitions',
+    'numCachedPartitions',
+    'memoryUsed',
+    'diskUsed'
+]
+
+
+class Spark(CollectorBase):
+    def __init__(self, config, logger, readq):
+        super(Spark, self).__init__(config, logger, readq)
+
+        self.spark_cluster_mode = self.get_config('spark_cluster_mode', SPARK_STANDALONE_MODE)
+        self.host = self.get_config('spark_host', "localhost")
+        self.port = self.get_config('spark_port', 8080)
+        self.spark_url = 'http://%s:%s' % (self.host, self.port)
+
+    def __call__(self):
+        spark_apps = self._get_running_apps()
+        self._spark_job_metrics(spark_apps)
+        self._spark_stage_metrics(spark_apps)
+        self._spark_executor_metrics(spark_apps)
+
+    def _get_running_apps(self):
```

Comment:
_get_running_apps is an expensive operations, we probably shouldn't run in each __call__